fenicsadapter/fenicsadapter.py

```diff
@@ -75,46 +75,28 @@ def __init__(self, adapter_config_filename='precice-adapter-config.json'):
         # Necessary bools for enforcing proper control flow / warnings to user
         self._first_advance_done = False
         self._apply_2d_3d_coupling = False
-        self._non_standard_initialization = False
 
-    def create_coupling_expression(self, data=None):
+    def create_coupling_expression(self):
         """
         Creates a FEniCS Expression in the form of an object of class GeneralInterpolationExpression or
         ExactInterpolationExpression. The adapter will hold this object till the coupling is on going.
 
-        Parameters
-        ----------
-        data : array_like
-            The coupling data. A numpy array [N x D] where N = number of vertices and D = dimensions of the data, i.e.
-            for scalar valued data D = 1 and for vector valued data D = dimension of problem.
-            If no data is provided then a numpy array of shape (N, D) with zero values is used.
-
         Returns
         -------
         coupling_expression : Object of class dolfin.functions.expression.Expression
             Reference to object of class GeneralInterpolationExpression or ExactInterpolationExpression.
         """
-        if self._non_standard_initialization and data is None:
-            warn("initialize_data() was called but the data was not supplied for creating the coupling expression")
+
+        if not (self._read_function_type is FunctionType.SCALAR or self._read_function_type is FunctionType.VECTOR):
+            raise Exception("No valid read_function is provided in initialization. Cannot create coupling expression")
 
         try:  # works with dolfin 1.6.0
             # element information must be provided, else DOLFIN assumes scalar function
             coupling_expression = self._my_expression(element=self._function_space.ufl_element())
         except (TypeError, KeyError):  # works with dolfin 2017.2.0
             coupling_expression = self._my_expression(element=self._function_space.ufl_element(), degree=0)
 
-        n_vertices, _ = self._coupling_mesh_vertices.shape
-
-        if data is None:
-            # standard initialization with zero valued data
-            if self._read_function_type is FunctionType.SCALAR:
-                data = np.zeros(n_vertices)
-            elif self._read_function_type is FunctionType.VECTOR:
-                data = np.zeros((n_vertices, self._fenics_dimensions))
-            else:
-                raise Exception("No valid read_function is provided in initialization. Cannot create coupling expression")
-
-        self.update_coupling_expression(coupling_expression, data)
+        coupling_expression._function_type = self._read_function_type
```

Comment:
Accessing (or creating) the member `coupling_expression._function_type` is a quick-and-dirty solution. We should better us a function call like `coupling_expression.set_function_type(self._read_function_type)` to respect the privacy of our `CouplingExpression`.